Update docs in FTP module (#247)
ballerina-platform/module-ballerina-ftp: ftp-ballerina/Module.md

## Overview

> This module provides an FTP client and an FTP server listener implementation to facilitate an FTP connection to a remote location.

Reviewer: ```suggestion
This module provides an FTP client and an FTP server listener implementation to facilitate an FTP connection connected to a remote location.
```

### FTP Client

The `ftp:Client` connects to an FTP server and performs various operations on the files. Currently, it supports the
generic FTP operations; `get`, `delete`, `put`, `append`, `mkdir`, `rmdir`, `isDirectory`, `rename`, `size`, and
 `list`.

> An FTP client endpoint is defined using the parameters `protocol` and `host`, and optionally the `port` and

Reviewer: ```suggestion
An FTP client endpoint is defined using the `protocol` and `host` parameters and optionally the `port` and
```

> `secureSocket`. Authentication configuration can be configured using the `secureSocket` parameter for basicAuth,

Reviewer: ```suggestion
`secureSocket`. Authentication configuration can be configured using the `secureSocket` parameter for Basic Auth,
```

private key, or TrustStore/Keystore.

##### Creating a Client

> The following code creates an FTP client and perform I/O operations, which connects to the FTP server with Basic Auth.

Reviewer: ```suggestion
The following code creates an FTP client and performs the I/O operations, which connect to the FTP server with Basic Auth.
```

```ballerina

> // Define FTP client configuration

Reviewer: ```suggestion
// Define the FTP client configuration.
```

ftp:ClientEndpointConfig ftpConfig = {
    protocol: ftp:FTP,
    host: "<The FTP host>",
    port: <The FTP port>,
    secureSocket: {
        basicAuth: {
            username: "<The FTP username>",
            password: "<The FTP passowrd>"
        }
    }
};

> // Create FTP client
> ftp:Client ftpClient = new(ftpConfig);

Reviewer: ```suggestion
// Create the FTP client.
ftp:Client ftpClient = new(ftpConfig);
```

```

##### Creating a Directory

The following code creates a directory in the remote FTP server.

```ballerina
ftp:Error? mkdirResponse = ftpClient->mkdir("<The directory path>");
```

##### Uploading File to a Remote Server

The following code uploads a file to a remote FTP server.

```ballerina
io:ReadableByteChannel summaryChannel
    = check io:openReadableFile("<The local data source path>");
ftp:Error? putResponse = ftpClient->put("<The resource path>", summaryChannel);
```

##### Compressing and Uploading a File to a Remote Server

The following code compresses and uploads a file to a remote FTP server.

```ballerina
io:ReadableByteChannel inputChannel
    = check io:openReadableFile("<Local data source path>");
// Set the optional boolean flag as 'true' to compress before uploading
ftp:Error? compressedPutResponse = ftpClient->put("<Resource path>",
    inputChannel, true);
```

##### Getting the Size of a Remote File

> The following code get and size of a file of a file in remote FTP server.

Reviewer: ```suggestion
The following code gets the size of a file in a remote FTP server.
```

```ballerina
int|ftp:Error sizeResponse = ftpClient->size("<The resource path>");
```

##### Reading the Content of a Remote File

> The following code read the content of a file in remote FTP server.

Reviewer: ```suggestion
The following code reads the content of a file in a remote FTP server.
```

```ballerina
io:ReadableByteChannel getResponse = check ftpClient->get("<The file path>");
io:ReadableCharacterChannel? characters
    = new io:ReadableCharacterChannel(getResponse, "utf-8");
if (characters is io:ReadableCharacterChannel) {
    string output = check characters.read(<No of characters to read>);
    var closeResult = characters.close();
}
```

##### Renaming or Moving a Remote file to Another Remote Location in the Same FTP Server

> The following rename or move remote a file to another location in the same remote FTP server.

Reviewer: ```suggestion
The following code renames or moves a file to another location in the same remote FTP server.
```

```ballerina
ftp:Error? renameResponse = ftpClient->rename("<The source file path>",
    "<The destination file path>");
```

##### Deleting a Remote File

> The following delete a remote file in a remote FTP server.

Reviewer: ```suggestion
The following code deletes a remote file in a remote FTP server.
```

> The following delete a remote file in a remote FTP server.

Reviewer: ```suggestion
The following code deletes a remote file in a remote FTP server.
```

```ballerina
ftp:Error? deleteResponse = ftpClient->delete("<The resource path>");
```

##### Removing a Directory From a Remote Server

> The following remove a directory in a remote FTP server.

Reviewer: ```suggestion
The following code removes a directory in a remote FTP server.
```

```ballerina
ftp:Error? rmdirResponse = ftpClient->rmdir("<The directory path>");
```

###FTP Listener

The `ftp:Listener` is used to listen to a remote FTP location and trigger a `WatchEvent` type of event when new
files are added to or deleted from the directory. The `fileResource` function is invoked when a new file is added
and/or deleted.

> An FTP listener endpoint is defined using the mandatory parameters `protocol`, `host` and  `path`. Authentication

Reviewer 2: ```suggestion
An FTP listener endpoint is defined using the mandatory parameters `protocol`, `host`, and  `path`. Authentication
```

> An FTP listener endpoint is defined using the mandatory parameters `protocol`, `host`, and  `path`. Authentication

Reviewer: ```suggestion
An FTP listener endpoint is defined using the mandatory `protocol`, `host`, and  `path` parameters. The authentication
```

> configuration can be done using `secureSocket` and polling interval can be configured using `pollingInterval`.

Reviewer: ```suggestion
configuration can be done using a `secureSocket` and the polling interval can be configured using the `pollingInterval` parameter.
```

> Default polling interval is 60 seconds.

Reviewer 2: ```suggestion
The default polling interval is 60 seconds.
```

The `fileNamePattern` parameter can be used to define the type of files the FTP listener endpoint will listen to.
For instance, if the listener gets invoked for text files, the value `(.*).txt` can be given for the config.

##### Creating a Listener

The FTP Listener can be used to listen to a remote directory. It will keep listening to the specified directory and

> periodically notify on file addition and deletion.

Reviewer: ```suggestion
notify on file addition and deletion periodically.
```

## Overview

This module provides an FTP client and an FTP server listener implementation to facilitate an FTP connection to a remote location.

> ###FTP Client

Reviewer 3: ```suggestion
### FTP Client
```

The `ftp:Client` connects to an FTP server and performs various operations on the files. Currently, it supports the
generic FTP operations; `get`, `delete`, `put`, `append`, `mkdir`, `rmdir`, `isDirectory`, `rename`, `size`, and
 `list`.

An FTP client endpoint is defined using the parameters `protocol` and `host`, and optionally the `port` and
`secureSocket`. Authentication configuration can be configured using the `secureSocket` parameter for basicAuth,
private key, or TrustStore/Keystore.

> ##### Creating a client

Reviewer 4: Need to use title case.

Author: Changed in all the similar instances.

The following code creates an FTP client and perform I/O operations, which connects to the FTP server with Basic Auth.
```ballerina
import ballerina/ftp;

> 

Reviewer 5: we don't need an import statement. wdyt?

Author: Yeah. Removed in all the import occurrences.

// Define FTP client configuration
ftp:ClientEndpointConfig ftpConfig = {
    protocol: ftp:FTP,
    host: "<The FTP host>",
    port: <The FTP port>,
    secureSocket: {
        basicAuth: {
            username: "<The FTP username>",
            password: "<The FTP passowrd>"
        }
    }
};

// Create FTP client
ftp:Client ftpClient = new(ftpConfig);
```

##### Creating a directory

The following code creates a directory in the remote FTP server.

```ballerina
ftp:Error? mkdirResponse = ftpClient->mkdir("<The directory path>");
if (mkdirResponse is ftp:Error) {
    log:printError("Error occured in creating directory", err = mkdirResponse);
    return;

> }

Reviewer 5: IMO, we don't need this error handling part.

Author: Removed in all the similar instances.

```

##### Uploading file to a remote server

The following code uploads a file to a remote FTP server.

```ballerina

> io:ReadableByteChannel|error summaryChannel

Reviewer 4: This section needs to be updated.

Author: Will update with new `stream` based APIs once the code is changed.

    = io:openReadableFile("<The local data source path>");
if(summaryChannel is io:ReadableByteChannel){
    ftp:Error? putResponse = ftpClient->put("<The resource path>", summaryChannel);
    if(putResponse is ftp:Error) {
        log:printError("Error occured in uploading content", err = putResponse);
        return;
    }
}
```

##### Compressing and uploading a file to a remote server

The following code compresses and uploads a file to a remote FTP server.

```ballerina
io:ReadableByteChannel|error inputChannel
    = io:openReadableFile("<Local data source path>");
if (inputChannel is io:ReadableByteChannel) {
    // Set the optional boolean flag as 'true' to compress before uploading
    ftp:Error? compressedPutResponse = ftpClient->put("<Resource path>",
        inputChannel, true);
    if (compressedPutResponse is ftp:Error) {
        log:printError("Error occured in uploading content",
            err = compressedPutResponse);
        return;
    }
}
```

##### Getting the size of a remote file

The following code get and size of a file of a file in remote FTP server.

```ballerina

> var sizeResponse = ftpClient->size("<The resource path>");

Reviewer 5: shall we replace the `var` type with actual return types?
We don't need result handling part.

Reviewer 5: please check other places

Author: Changed all the relevant `var` type definition occurrences.
